catalog/plugin/catalog-plugin-duplication/src/main/java/ddf/catalog/metacard/duplication/DuplicationPlugin.java

```diff
@@ -0,0 +1,270 @@
+import ddf.catalog.operation.SourceResponse;
+import ddf.catalog.operation.UpdateRequest;
+import ddf.catalog.operation.impl.QueryImpl;
+import ddf.catalog.operation.impl.QueryRequestImpl;
+import ddf.catalog.plugin.PluginExecutionException;
+import ddf.catalog.plugin.PreIngestPlugin;
+import ddf.catalog.plugin.StopProcessingException;
+import ddf.catalog.source.SourceUnavailableException;
+import ddf.catalog.source.UnsupportedQueryException;
+
+public class DuplicationPlugin implements PreIngestPlugin {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DuplicationPlugin.class);
+
+    private final CatalogFramework catalogFramework;
+
+    private final FilterBuilder filterBuilder;
+
+    private String[] failOnDuplicateList;
+
+    private String[] errorOnDuplicateList;
+
+    private String[] warnOnDuplicateList;
+
+    public DuplicationPlugin(CatalogFramework catalogFramework, FilterBuilder filterBuilder) {
+        this.catalogFramework = catalogFramework;
+        this.filterBuilder = filterBuilder;
+    }
+
+    /**
+     * Setter for the list of attributes to test for duplication in the local catalog.  Resulting
+     * attributes will cause an {@link StopProcessingException}
+     *
+     * @param attributeStrings
+     */
+    public void setFailOnDuplicateList(String[] attributeStrings) {
+        if (attributeStrings != null) {
+            this.failOnDuplicateList = Arrays.copyOf(attributeStrings, attributeStrings.length);
+        }
+    }
+
+    /**
+     * Setter for the list of attributes to test for duplication in the local catalog.  Resulting
+     * attributes will cause the {@link ddf.catalog.data.impl.BasicTypes#VALIDATION_ERRORS} attribute
+     * to be set on the metacard.
+     *
+     * @param attributeStrings
+     */
+    public void setErrorOnDuplicateList(String[] attributeStrings) {
+        if (attributeStrings != null) {
+            this.errorOnDuplicateList = Arrays.copyOf(attributeStrings, attributeStrings.length);
+        }
+    }
+
+    /**
+     * Setter for the list of attributes to test for duplication in the local catalog.  Resulting
+     * attributes will cause the {@link ddf.catalog.data.impl.BasicTypes#VALIDATION_WARNINGS} attribute
+     * to be set on the metacard.
+     *
+     * @param attributeStrings
+     */
+    public void setWarnOnDuplicateList(String[] attributeStrings) {
+        if (attributeStrings != null) {
+            this.warnOnDuplicateList = Arrays.copyOf(attributeStrings, attributeStrings.length);
+        }
+    }
+
+    @Override
+    public CreateRequest process(CreateRequest input)
+            throws PluginExecutionException, StopProcessingException {
+
+        if (input != null && input.getMetacards() != null) {
+            checkForDuplicates(input.getMetacards());
+        }
+
+        return input;
+    }
+
+    @Override
+    public UpdateRequest process(UpdateRequest input) throws PluginExecutionException {
+        return input;
```

Comment:
I was on the fence here.  It seems possible but an unusual scenario.  If a metacard (and possible related content) already exists, and a client updates it, could it become a duplicate of something in the catalog?  I suppose it could happen but I was a little concerned implementing this.  Once we have a card in our catalog it seems strange to then reject an update (or invalidate it) due to a possible duplicate scenario.